### DISCUSSION TITLE
Passing session to Quirrel queue

### DISCUSSION THREAD

**Discussion by leo-guinan** [blitz-js/blitz#2641] [Q&A] [answered]
Here's the workflow I'm trying to implement:

1. User creates an account - this works fine
2. User links Twitter account - this also works fine
3. Once the user links their account, I want to add a job to a Quirrel queue to populate initial data from their Twitter account. - this is where the problem is

I need to pass in the context to the Quirrel queue in order to save records to my database. But when I try, I got the error "TypeError: Converting circular structure to JSON".

It seems like calling Object.stringify on the context object that I'm passing into the queue has a circular reference in it.

Does anyone have any thoughts? Do I need to use something other than the mutator that requires the session context?

**Comment by flybayer** [comment]
cc @Skn0tt

**Comment by MrLeebo** [comment, marked as answer]
You didn't describe the context object you were trying to serialize, but I think you'd be better off specifying the exact attributes from the session data that the job must have in order to run rather than trying to serialize everything. Not only is that going to serialize a lot more data than you need and give you circular reference errors like this, but even if you deleted all of the circular keys you identified from the object today, there's no guarantee that a new circular key wouldn't get added to it tomorrow.

If you pick and choose the input parameters that you send to the job, then you can avoid this entire category of errors.

**Reply by leo-guinan** [reply]
Ok, that makes a lot of sense. 

So it sounds like I shouldn't try to use the mutator to do the db insert because that requires me to pass the context along from the request (I believe to be able to ensure that the user has the correct permissions). 

I'll have to take a look at how I can do the db insert a different way. 

Thanks!

**Reply by MrLeebo** [reply]
It's good to create a service function that performs the DB action with just the basic parameters and no HTTP context and call that function from both places. Technically, you can make the mutator's context nullable, but doing that just makes your design messy. It's sort of like leaving a sharp knife out and asking a toddler PLEASE don't pick it up. It's better to not leave knives around toddlers, and it's better to not try to access HTTP context from background workers.

**Comment by Skn0tt** [comment]
Hi, Quirrel guy here 👋 2nd what @MrLeebo says. In order to persist a payload in the the queue, Quirrel needs to serialize it. The context object is not serializable. You can, however, extract the important information from the context object, for example the ID of the requesting user, and put only that into the payload.

**Comment by Skn0tt** [comment]
You could extract your DB insertion logic into a function and call that from both your mutation and the Queue.
